aws-s3-private-bucket/main.tf

```diff
@@ -18,26 +18,54 @@ resource "aws_s3_bucket" "bucket" {
     enabled = var.enable_versioning
   }
 
-  lifecycle_rule {
-    enabled = true
+  dynamic "lifecycle_rule" {
+    for_each = var.lifecycle_rule
+
+    content {
+      id                                     = lookup(lifecycle_rule.value, "id", null)
```

Comment:
I think you can "just" do `lifecycle_rule.value.id`, right? If that's not the case then we should leave a comment about why we need to do this.

Reply:
That works, but not all `lifecycle_rule` properties are present in the default rule (in `variables.tf`). Thus if I did `lifecycle_rule.value.prefix`, Terraform would complain that it couldn't find it, hence the `lookup()` w/ the default values. 

I can definitely do the short-hand for the properties that are present in the default rule and keep the `lookup()` for those that aren't present, but it feels very un-uniform. Also, it's possible the default rule might change in the future - we wouldn't want to couple the two together. Thoughts?

Reply:
Yeah I see that now (sorry this is a new feature and I don't know all the subtleties yet).

Given that constraint I think it is fine to leave as-is until we find a more elegant way to do it.

Reply:
> for those that aren't present, but it feels very un-uniform.

Personally, I think that would be better. It signals intent clearly to the reader whether the programmer expects this value to exist, or that it might not.

> we wouldn't want to couple the two together. Thoughts?

Then we would also change its callsites to reflect the new signature?

Reply:
> whether the programmer expects this value to exist, or that it might not.

For most of these values, their existence depends on the configuration that you're trying to use right? IMO the default configuration provided in `variables.tf` shouldn't influence how it's being written here - what if we have add multiple default lifecycle_rules? 

I think signal to the programmer can be achieved via running `make plan` right?